kotlinx-coroutines-core/common/test/flow/operators/FlowOnTest.kt

```diff
@@ -314,4 +314,19 @@ class FlowOnTest : TestBase() {
             assertEquals(expected, value)
         }
     }
+
+    @Test
+    fun testCancelledFlowOn() = runTest {
+        assertFailsWith<CancellationException> {
```

Comment:
Thinking about the original issue (#1265), I am concerned about the fact that other channel-related operators may expose unexpected behaviour as well (aka touching view after its destruction).
E.g. the similar test with `combine` operator will execute its `collect` block.

I'd say we should re-visit our channel-related operators, measure the performance impact of checking cancellation in `collect` and either do it, or do it only in channel-related operators (and, maybe, provide a `checkedCollect` or something like this)

Reply:
It's not about checking in `collect`. It is about checking _on resume after suspension_. All our suspending function do check for cancellation on resume with exception of `Mutex.lock` , `Semaphore.acquire`, and `Channel.send/receive` function that do not do so, we need to revisit operation that use those functions (`flowOn` was one of them). In fact, we should really revisit atomicity of those underlying operations in the first place.

Reply:
Do you want to do it separately or as part of this PR?
E.g. the following code:
```
coroutineScope {
    val flow =  flow {
        emit(Unit) // emit to buffer
        cancel() // now cancel
    }

    flow.combine(flow) { u, _ -> u }.collect {
        // should not be reached, because cancelled before it runs
        expectUnreached()
    }
}
```

will throw `ISE(Should not be reached)`

Reply:
I'll fix it all as a part of this PR.